registry/storage/driver/swift/swift.go

```diff
@@ -0,0 +1,547 @@
+	headers["Range"] = "bytes=" + strconv.FormatInt(offset, 10) + "-"
+
+	file, _, err := d.Conn.ObjectOpen(d.Container, d.swiftPath(path), false, headers)
+
+	if err != nil {
+		if swiftErr, ok := err.(*swift.Error); ok && swiftErr.StatusCode == 416 {
+			return ioutil.NopCloser(bytes.NewReader(nil)), nil
+		}
+
+		return nil, parseError(path, err)
+	}
+
+	return file, nil
+}
+
+// WriteStream stores the contents of the provided io.Reader at a
+// location designated by the given path. The driver will know it has
+// received the full contents when the reader returns io.EOF. The number
+// of successfully READ bytes will be returned, even if an error is
+// returned. May be used to resume writing a stream by providing a nonzero
+// offset. Offsets past the current size will write from the position
+// beyond the end of the file.
+func (d *driver) WriteStream(ctx context.Context, path string, offset int64, reader io.Reader) (int64, error) {
+	var (
+		segments      []swift.Object
+		multi         io.Reader
+		paddingReader io.Reader
+		currentLength int64
+		cursor        int64
+	)
+
+	partNumber := 1
+	chunkSize := int64(d.ChunkSize)
+	zeroBuf := make([]byte, d.ChunkSize)
+
+	getSegment := func() string {
+		return d.swiftSegmentPath(path) + "/" + fmt.Sprintf("%016d", partNumber)
+	}
+
+	max := func(a int64, b int64) int64 {
+		if a > b {
+			return a
+		}
+		return b
+	}
+
+	info, _, err := d.Conn.Object(d.Container, d.swiftPath(path))
+	if err != nil {
+		if err == swift.ContainerNotFound || err == swift.ObjectNotFound {
+			// Create a object manifest
+			if err := d.createParentFolders(path); err != nil {
+				return 0, err
+			}
+			manifest, err := d.createManifest(path)
+			if err != nil {
+				return 0, parseError(path, err)
+			}
+			manifest.Close()
+		} else {
+			return 0, parseError(path, err)
+		}
+	} else {
+		// The manifest already exists. Get all the segments
+		currentLength = info.Bytes
+		segments, err = d.getAllSegments(path)
+		if err != nil {
+			return 0, parseError(path, err)
+		}
+	}
+
+	// First, we skip the existing segments that are not modified by this call
+	for i := range segments {
+		if offset < cursor+segments[i].Bytes {
+			break
+		}
+		cursor += segments[i].Bytes
+		partNumber++
+	}
+
+	// We reached the end of the file but we haven't reached 'offset' yet
+	// Therefore we add blocks of zeros
+	if offset >= currentLength {
+		for offset-currentLength >= chunkSize {
+			// Insert a block a zero
+			d.Conn.ObjectPut(d.Container, getSegment(),
+				bytes.NewReader(zeroBuf), false, "",
+				d.getContentType(), nil)
+			currentLength += chunkSize
+			partNumber++
+		}
+
+		cursor = currentLength
+		paddingReader = bytes.NewReader(zeroBuf)
+	} else if offset-cursor > 0 {
+		// Offset is inside the current segment : we need to read the
+		// data from the beginning of the segment to offset
+		file, _, err := d.Conn.ObjectOpen(d.Container, getSegment(), false, nil)
+		defer file.Close()
```

Comment:
``err``` must be checked before posting ```defer file.Close()```. As far as I understand from ```Connection.ObjectOpen``` source, it could return a nil pointer, and calling Close on nil leads to panic.